### PR TITLE
Unit convert

### PR DESCRIPTION
**Pull request by paxblueribbon**
Added functionality for converting units between metric and imperial, ease of human readability, and use of multiple units where applicable.

### CODE REVIEW

File: data/org.gnome.Recipes.gschema.xml [hunk 1]
```diff
@@ -95,7 +101,20 @@
         The setting determines whether the surprise dialog should be shown on startup.
       </description>
     </key>
-
+    <key name="volume-unit" enum="org.gnome.recipes.MetricOrImperial">
+      <default>'locale'</default>
+      <summary>The setting for which unit temperatures should be displayed in. </summary>
+      <description>
+         The setting for which unit temperatures should be displayed in. Default is 'locale',
```
Comment (matthiasclasen) [inline data/org.gnome.Recipes.gschema.xml]: This is about volumes, not temperatures. Copy-paste error

File: data/org.gnome.Recipes.gschema.xml [hunk 2]
```diff
@@ -95,7 +101,20 @@
...
+      </description>
+     </key>
+    <key name="weight-unit" enum="org.gnome.recipes.MetricOrImperial">
+      <default>'locale'</default>
+      <summary>The setting for which unit temperatures should be displayed in. </summary>
+      <description>
+         The setting for which unit temperatures should be displayed in. Default is 'locale',
```
Comment (matthiasclasen) [inline data/org.gnome.Recipes.gschema.xml]: Mass, not temperature.

File: src/gr-convert-units.c [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+                fmt = nl_langinfo (_NL_MEASUREMENT_MEASUREMENT);
+                
+                if (fmt && *fmt == 2)
+                    unit = GR_PREFERRED_UNIT_IMPERIAL;
+                else
+            #endif
+                    unit = GR_PREFERRED_UNIT_METRIC;
+        }
+        return unit;
+}
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: Looking at these 3 functions, it strikes me that this is essentially the same code. The only difference is the setting name. Maybe we should turn this into a little helper function that takes the setting as an extra argument: 

GrPreferredUnit
gr_convert_get_metric_or_imperial (const char *setting)
{
...
}

and then

GrPreferredUnit
gr_convert_weight_unit (void)
{
  return gr_convert_get_metric_or_imperial ("weight-unit");
}

File: src/gr-convert-units.c [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+        else if (unit1 == GR_TEMPERATURE_UNIT_FAHRENHEIT &&
+            user_unit == GR_TEMPERATURE_UNIT_CELSIUS) {
+                num1 = (num1 - 32) / 1.8;
+                unit1 = user_unit;
+            }
+                                
+            *unit = unit1;
+            *num = num1;
+}
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: indentation is messy here.

File: src/gr-ingredients-viewer.c [hunk 5]
```diff
@@ -316,24 +318,49 @@ gr_ingredients_viewer_set_ingredients (GrIngredientsViewer *viewer,
...
+                
+                double amount;
+                GrUnit unit;
                 GtkWidget *row;
+                GrDimension dimension;
+                double scale = viewer->scale;
+
+                unit = gr_ingredients_list_get_unit(ingredients, ings[i]);
+                amount = gr_ingredients_list_get_amount(ingredients, viewer->title, ings[i]) * scale;
+                dimension = gr_unit_get_dimension(unit);
```
Comment (matthiasclasen) [inline src/gr-ingredients-viewer.c]: formatting details: please put a space before (

File: src/gr-ingredients-viewer.c [hunk 6]
```diff
@@ -316,24 +318,49 @@ gr_ingredients_viewer_set_ingredients (GrIngredientsViewer *viewer,
...
+
+                if (dimension) {
+                
+                        if (dimension == GR_DIMENSION_VOLUME) {
+                                GrPreferredUnit user_volume_unit = gr_convert_get_volume_unit();
+                                gr_convert_volume(&amount, &unit, user_volume_unit);
```
Comment (matthiasclasen) [inline src/gr-ingredients-viewer.c]: here too, and other places

File: src/gr-convert-units.c [hunk 7]
```diff
@@ -322,4 +322,111 @@ gr_convert_weight (double *amount, GrUnit *unit, GrPreferredUnit user_weight_uni
...
+                ;
+    }
+
+                if (*unit == unit1) {
+                unit_changed = FALSE; 
+                }
+                
+                *amount = amount1;
+                *unit = unit1;
+    }
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: looks like a mix between 4, 8 and 0 space indent here...

File: src/gr-ingredients-viewer.c [hunk 8]
```diff
@@ -342,6 +342,8 @@ gr_ingredients_viewer_set_ingredients (GrIngredientsViewer *viewer,
                         }
                 }
 
+                gr_convert_human_readable(&amount, &unit);
```
Comment (matthiasclasen) [inline src/gr-ingredients-viewer.c]: space before (, please

File: src/gr-convert-units.c [hunk 9]
```diff
@@ -422,11 +422,129 @@ gr_convert_human_readable (double *amount, GrUnit *unit)
...
+    
+    *amount = amount1;
+    *unit = unit1;
+    }
+}
+
+void
+gr_convert_multiple_units (double *amount1, GrUnit *unit1, double *amount2, GrUnit *unit2)
+{
+
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: extraneous empty line here

File: src/gr-convert-units.c [hunk 10]
```diff
@@ -422,11 +422,129 @@ gr_convert_human_readable (double *amount, GrUnit *unit)
...
+    GrUnit u2 = *unit2;
+    double fractional, integer;
+
+    gr_convert_human_readable(&n1, &u1);
+
+    fractional = modf(n1, &integer);
+
+    if (u1 != GR_UNIT_UNKNOWN) {
+            if (fractional > 0)
+            {
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: { goes on the same line as the if, please

File: src/gr-convert-units.c [hunk 11]
```diff
@@ -422,11 +422,129 @@ gr_convert_human_readable (double *amount, GrUnit *unit)
...
+                n2 = fractional;
+                u2 = u1;
+                gr_convert_human_readable(&n2, &u2);
+
+                if (u1 != u2)
+                {
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: here too, and elsewhere

File: src/gr-convert-units.c [hunk 12]
```diff
@@ -422,11 +422,129 @@ gr_convert_human_readable (double *amount, GrUnit *unit)
...
+    *unit1 = u1;
+    *amount2 = n2;
+    *unit2 = u2;
+}
+
+void
+gr_convert_format_for_display  (GString *s, double a1, GrUnit u1, double a2, GrUnit u2) 
+{
+
+    if (u1 == GR_UNIT_UNKNOWN || (!u1)) {
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: the !ui check doesn't really make sense.
Effectively, that is a check for u1 == 0, which is the same as u1 == GR_UNIT_UNKNOWN

File: src/gr-convert-units.c [hunk 13]
```diff
@@ -422,11 +422,129 @@ gr_convert_human_readable (double *amount, GrUnit *unit)
...
+        g_string_append (s, gr_unit_get_display_name (u1));
+    }
+
+    else {
+        g_autofree char *num1 = NULL;
+        g_autofree char *num2 = NULL;
+
+        num1 = gr_number_format (a1);
+        num2 = gr_number_format (a2);
+        g_string_append(s, num1);
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: space before (

File: src/gr-convert-units.c [hunk 14]
```diff
@@ -422,11 +422,129 @@ gr_convert_human_readable (double *amount, GrUnit *unit)
...
+        g_string_append(s, " ");
+        g_string_append(s, gr_unit_get_display_name (u1));
```
Comment (matthiasclasen) [inline src/gr-convert-units.c]: I think we decided to stick with abbreviations, instead of display names ?

File: src/gr-ingredients-list.c [hunk 15]
```diff
@@ -279,16 +279,14 @@ gr_ingredients_list_get_unit (GrIngredientsList *ingredients,
...
                                 const char        *name)
 {
         GList *l;
 
         for (l = ingredients->ingredients; l; l = l->next) {
                 Ingredient *ing = (Ingredient *)l->data;
 
-                if (g_strcmp0 (segment, ing->segment) == 0 &&
-                    g_strcmp0 (name, ing->name) == 0) {
+                if (g_strcmp0 (name, ing->name) == 0) {
```
Comment (matthiasclasen) [inline src/gr-ingredients-list.c]: This change looks unrelated, and ... wrong. Why do we stop comparing the segment here ?

File: src/gr-ingredients-viewer.c [hunk 16]
```diff
@@ -322,43 +322,17 @@ gr_ingredients_viewer_set_ingredients (GrIngredientsViewer *viewer,
...
-                else {
-                        snprintf(for_display, sizeof for_display, "%s %s", a_final, u_final);
-                }
-
-                u_final = for_display;
+                gr_convert_format(s, amount, unit);
 
                 row = g_object_new (GR_TYPE_INGREDIENTS_VIEWER_ROW,
-                                        "unit", u_final,
+                                        "unit", g_strdup (s->str),
```
Comment (matthiasclasen) [inline src/gr-ingredients-viewer.c]: This is going to leak a copy of the string. Passing s->str should be fine, g_object_new is already going make a copy of the string for you.

File: src/gr-recipe-printer.c [hunk 17]
```diff
@@ -193,16 +193,19 @@ begin_print (GtkPrintOperation *operation,
...
 
-        for (j = 0; segs[j]; j++) {
-                ings = gr_ingredients_list_get_ingredients (ingredients, segs[j]);
-                for (i = 0; ings[i]; i++) {
-                        g_autofree char *unit = NULL;
-
-                        unit = gr_ingredients_list_scale_unit (ingredients, segs[j], ings[i], 1.0);
-                        g_string_append (s, unit);
-                        g_string_append (s, " \n");
-                }
```
Comment (matthiasclasen) [inline src/gr-recipe-printer.c]: the segments get lost here, and I don't really know why that should be ok.

File: src/gr-unit.c [hunk 18]
```diff
@@ -16,152 +16,154 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#include "config.h"
```
Comment (matthiasclasen) [inline src/gr-unit.c]: What happened here ? change of line endings ? the whole file seems to have been replaced without change ?

File: src/gr-unit.h [hunk 19]
```diff
@@ -16,55 +16,56 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#pragma once
```
Comment (matthiasclasen) [inline src/gr-unit.h]: Same here.

File: src/gr-recipe-printer.c [hunk 20]
```diff
@@ -32,6 +32,7 @@
 #include "gr-meal.h"
 #include "gr-season.h"
 #include "gr-utils.h"
+#include "gr-convert-units.h"
```
Comment (matthiasclasen) [inline src/gr-recipe-printer.c]: This looks just like a fixup to the previous commit

**Comment by gnomesysadmins** [comment]
Thank you for contributing to recipes!

recipes uses Bugzilla for code review.

If you have never contributed to GNOME before make sure you have read the
getting started documentation:
http://www.gnome.org/get-involved

Otherwise please visit
https://wiki.gnome.org/Newcomers
and follow the instructions there to upload your change to Bugzilla.
